samza-core/src/main/java/org/apache/samza/runtime/ContainerLaunchUtil.java

```diff
@@ -132,6 +136,10 @@ private static void run(
 
       container.setContainerListener(
           new SamzaContainerListener() {
```

Comment:
Could we add some unit tests for this? This anonymous class is getting pretty complex. Maybe this is also a good time to pull it out into its own class.
You'll have to be a little careful since you don't actually want to add a real shutdown hook to the test process though. One way you could do it is to pull `Runtime.getRuntime().addShutdownHook` and `Runtime.getRuntime().removeShutdownHook` into internal helper methods and then use a spy to mock those internal methods. And then maybe you could capture the shutdown hook thread using Mockito and call it directly when necessary.